File: 백준/Gold/5639. 이진 검색 트리/이진 검색 트리.py

from dataclasses import dataclass
from typing import Optional
import itertools, sys

input = sys.stdin.readline


def solve(seq:list[int]):
    '''
    Args:
        seq: list of numbers of pre-order tree tranversal.
    Returns:
        re-write postorder traversed result into seq list.
    '''
    @dataclass
    class Node:
        data: int = 0
        left: Optional["Node"] = None
        right: Optional["Node"] = None

    def add_preorder(tree, data):
        assert data != tree.data
        # going deeper down..
        node = tree
        while True:
            if data < node.data:
                if not node.left:
                    node.left = Node(data)
                    return
                node = node.left
            else: # node.data < data
                if not node.right:
                    node.right = Node(data)
                    return
                node = node.right

    assert len(seq) > 0 # should not empty
    root = Node(seq[0]) # consume first element
    for d in itertools.islice(seq, 1, None): # from second..
        add_preorder(root, d)

    idx_out = 0
    def write_postorder(tree):
        nonlocal idx_out
        stack:list[Node|int] = [tree]
            # list element: node or integer
        while stack:
            node = stack.pop()
            if isinstance(node, int): # this is not Node, but the number to be output.
                seq[idx_out] = node
                idx_out += 1
                continue
            stack.append(node.data) # append as 'integer' value

            if node.right:
                stack.append(node.right)
            if node.left:
                stack.append(node.left)
        return
    write_postorder(root)
    return

#--------
seq = []
while True:
    ln = input().strip()
    if not ln: break  # EOF
    seq.append(int(ln))

solve(seq)
for d in seq:
    print(d)

File: 백준/Gold/5639. 이진 검색 트리/이진 검색 트리.cc

#include <iostream>

using namespace std;

struct Node {
    int d;
    Node *lf, *rg; // left, right child

    Node(int d, Node* p=nullptr) {
        this->d = d;
        lf = rg = nullptr;
    }
};

void add_node(Node* tree, int d)
{
    if (d < tree->d) {
        if (tree->lf)
            add_node(tree->lf, d);
        else
            tree->lf = new Node(d);
    }
    else {
        if (tree->rg)
            add_node(tree->rg, d);
        else
            tree->rg = new Node(d);
    }
}

void print_postorder(Node* tree)
{
    if (tree->lf) print_postorder(tree->lf);
    if (tree->rg) print_postorder(tree->rg);
    printf("%d\n", tree->d);
}
void del_tree(Node* tree)
{
    if (!tree) return;
    if (tree->lf) del_tree(tree->lf);
    if (tree->rg) del_tree(tree->rg);
    tree->lf = tree->rg = nullptr;
    delete tree;
}

int main()
{
    int n;
    Node* root = nullptr;

    while (scanf("%d", &n) != EOF) {
        if (root)
            add_node(root, n);
        else
            root = new Node(n);
    }
    // print using post order
    print_postorder(root);

    del_tree(root);
    return 0;
}
